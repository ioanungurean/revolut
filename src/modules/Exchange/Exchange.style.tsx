import styled from "styled-components";

export const StyledExchange = styled.div`
  display: flex;
  flex-direction: column;
`;

export const StyledHeader = styled.div`
  font-size: 24px;
  font-weight: bold;
  display: flex;
`;

export const StyledSubHeader = styled.div`
  font-size: 12px;
  display: flex;
  margin-bottom: 12px;
  color: #2684ff;
`;

export const StyledCurrency = styled.div`
  display: flex;
  padding: 12px;
  flex-direction: column;
  width: 360px;
`;

export const StyledContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
`;

export const StyledInput = styled.input`
  text-align: right;
  width: 140px;
  padding: 4px;
`;

export const StyledBalance = styled.span`
  font-size: 12px;
  padding-top: 6px;
`;

export const StyledErrorMessage = styled.span`
  font-size: 12px;
  padding-top: 6px;
  color: red;
`;

export const StyledButton = styled.button`
  font-size: 16px;

  width: 360px;
  padding: 12px 36px;

  cursor: pointer;
  text-align: center;

  color: white;
  border: none;
  background-color: #2684ff;

  &:hover {
    background-color: #0c50a6;
  }

  &:disabled {
    cursor: default;
    background-color: #c8c8c8;
  }
`;
