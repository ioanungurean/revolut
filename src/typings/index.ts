export enum ECurrency {
  Usd = "USD",
  Eur = "EUR",
  Gbp = "GBP",
  Aud = "AUD",
  Cad = "CAD",
}

export enum EActionType {
  Exchange = "EXCHANGE",
  ChangeFrom = "CHANGE_FROM",
  ChangeTo = "CHANGE_TO",
}

export interface IAccount {
  id: ECurrency;
  balance: string;
}

export interface ICurrency {
  id: ECurrency;
  amount: string;
}

export type TState = {
  accounts: IAccount[];
  currencyFrom: ICurrency;
  currencyTo: ICurrency;
};

export type TActionPayload = {
  currency: ICurrency;
};

export type TAction =
  | { type: EActionType.Exchange }
  | {
      type: EActionType.ChangeFrom;
      payload: TActionPayload["currency"];
    }
  | { type: EActionType.ChangeTo; payload: TActionPayload["currency"] };
