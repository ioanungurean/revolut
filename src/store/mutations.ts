import currency from "currency.js";
import { TState, TActionPayload } from "../typings";

export const exchange = (state: TState): TState => {
  const { accounts, currencyFrom, currencyTo } = state;

  const updatedAccounts = accounts.map((account) => {
    if (account.id === currencyFrom.id) {
      return {
        ...account,
        balance: `${
          currency(account.balance).subtract(currencyFrom.amount).value
        }`,
      };
    }

    if (account.id === currencyTo.id) {
      return {
        ...account,
        balance: `${currency(account.balance).add(currencyTo.amount).value}`,
      };
    }

    return account;
  });

  return {
    ...state,
    currencyFrom: {
      ...state.currencyFrom,
      amount: "",
    },
    currencyTo: {
      ...state.currencyTo,
      amount: "",
    },
    accounts: updatedAccounts,
  };
};

export const setFrom = (
  state: TState,
  payload: TActionPayload["currency"]
): TState => {
  return {
    ...state,
    currencyFrom: payload,
  };
};

export const setTo = (
  state: TState,
  payload: TActionPayload["currency"]
): TState => {
  return {
    ...state,
    currencyTo: payload,
  };
};
