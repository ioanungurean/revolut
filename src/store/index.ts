import { exchange, setFrom, setTo } from "./mutations";

import { ECurrency, EActionType, TAction, TState } from "../typings";

export const initialState: TState = {
  accounts: [
    {
      id: ECurrency.Usd,
      balance: "874.12",
    },
    {
      id: ECurrency.Eur,
      balance: "45.04",
    },
    {
      id: ECurrency.Gbp,
      balance: "1000.00",
    },
    {
      id: ECurrency.Aud,
      balance: "3232.12",
    },
    {
      id: ECurrency.Cad,
      balance: "453.12",
    },
  ],
  currencyFrom: { id: ECurrency.Usd, amount: "" },
  currencyTo: { id: ECurrency.Eur, amount: "" },
};

export const reducer = (state: TState, action: TAction) => {
  switch (action.type) {
    case EActionType.Exchange:
      return exchange(state);
    case EActionType.ChangeFrom:
      return setFrom(state, action.payload);
    case EActionType.ChangeTo:
      return setTo(state, action.payload);
    default:
      throw new Error();
  }
};
