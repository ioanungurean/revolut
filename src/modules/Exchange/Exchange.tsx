import React, { useEffect, useReducer } from "react";
import currency from "currency.js";

import { initialState, reducer } from "../../store";
import { useFetchRates } from "../../hooks/useFetchRates";
import { checkIfValidNumber } from "../../utils";

import Select from "../../components/Select";

import { ECurrency, EActionType, IAccount } from "../../typings";

import {
  StyledExchange,
  StyledHeader,
  StyledSubHeader,
  StyledCurrency,
  StyledContainer,
  StyledInput,
  StyledBalance,
  StyledErrorMessage,
  StyledButton,
} from "./Exchange.style";

export interface ExchangeProps {}

const Exchange: React.FC<ExchangeProps> = () => {
  const [{ accounts, currencyFrom, currencyTo }, dispatch] = useReducer(
    reducer,
    initialState
  );

  const { data } = useFetchRates(currencyFrom, currencyTo);

  useEffect(() => {
    if (data) {
      const amount = currencyFrom.amount
        ? `${currency(currencyFrom.amount).multiply(data).value}`
        : "";

      dispatch({
        type: EActionType.ChangeTo,
        payload: {
          id: currencyTo.id,
          amount,
        },
      });
    }
  }, [data, currencyFrom.amount]);

  const currencyOptions: ECurrency[] = accounts.map(({ id }) => id);

  const accountFrom = accounts.find(
    ({ id }) => id === currencyFrom.id
  ) as IAccount;
  const accountTo = accounts.find(({ id }) => id === currencyTo.id) as IAccount;

  const handleSelectChange = (
    id: ECurrency,
    type: EActionType,
    amount: string
  ) => {
    dispatch({
      type,
      payload: {
        id,
        amount,
      },
    });
  };

  const handleInputChange = (
    ev: React.ChangeEvent<HTMLInputElement>,
    type: EActionType,
    id: ECurrency
  ) => {
    if (checkIfValidNumber(ev.target.value)) {
      dispatch({
        type,
        payload: {
          id,
          amount: ev.target.value,
        },
      });
    }
  };

  return (
    <StyledExchange>
      <StyledHeader>Sell {accountFrom.id}</StyledHeader>

      <StyledSubHeader>
        Market order â€¢ 1 {accountFrom.id} = {data} {accountTo.id}
      </StyledSubHeader>

      <StyledCurrency>
        <StyledContainer>
          <Select
            id="select-currency-from"
            options={currencyOptions}
            defaultValue={accountFrom.id}
            onChange={(selectedOption: ECurrency) =>
              handleSelectChange(
                selectedOption,
                EActionType.ChangeFrom,
                currencyFrom.amount
              )
            }
          />
          <StyledInput
            data-testid="input-amount-from"
            value={currencyFrom.amount}
            placeholder={`0.00 ${currencyFrom.id}`}
            onChange={(ev: React.ChangeEvent<HTMLInputElement>) =>
              handleInputChange(ev, EActionType.ChangeFrom, accountFrom.id)
            }
          />
        </StyledContainer>
        <StyledContainer>
          <StyledBalance data-testid="account-balance-from">
            Balance: {accountFrom.balance} {accountFrom.id}
          </StyledBalance>
          {currency(currencyFrom.amount).value >
            currency(accountFrom.balance).value && (
            <StyledErrorMessage>exceed balance</StyledErrorMessage>
          )}
        </StyledContainer>
      </StyledCurrency>

      <StyledCurrency>
        <StyledContainer>
          <Select
            id="select-currency-to"
            options={currencyOptions}
            defaultValue={accountTo.id}
            onChange={(selectedOption: ECurrency) =>
              handleSelectChange(
                selectedOption,
                EActionType.ChangeTo,
                currencyTo.amount
              )
            }
          />
          <StyledInput
            data-testid="input-amount-to"
            value={currencyTo.amount}
            placeholder={`0.00 ${currencyTo.id}`}
            onChange={(ev: React.ChangeEvent<HTMLInputElement>) =>
              handleInputChange(ev, EActionType.ChangeTo, accountTo.id)
            }
          />
        </StyledContainer>
        <StyledContainer>
          <StyledBalance data-testid="account-balance-to">
            Balance: {accountTo.balance} {accountTo.id}
          </StyledBalance>
        </StyledContainer>
      </StyledCurrency>

      <StyledButton
        data-testid="sell-button"
        disabled={
          !data ||
          !currency(currencyFrom.amount).value ||
          accountFrom.id === accountTo.id ||
          currency(currencyFrom.amount).value >
            currency(accountFrom.balance).value
        }
        onClick={() =>
          dispatch({
            type: EActionType.Exchange,
          })
        }
      >
        Sell {accountFrom.id} to {accountTo.id}
      </StyledButton>
    </StyledExchange>
  );
};

export default Exchange;
