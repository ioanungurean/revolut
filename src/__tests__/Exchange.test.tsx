import React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import currency from "currency.js";
import { render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FakeTimers from "@sinonjs/fake-timers";
import fetchMock from "fetch-mock-jest";

import { initialState } from "../store";

import { ECurrency, IAccount } from "../typings";

import Exchange from "../modules/Exchange";

const fetchOptions = {
  method: "GET",
  headers: {
    "x-rapidapi-host": "currency-exchange.p.rapidapi.com",
    "x-rapidapi-key": "075300c478msh0e398192b31d907p1ca169jsn375580d6aab4",
  },
};

const getFetchUrl = (from: ECurrency, to: ECurrency) =>
  `https://currency-exchange.p.rapidapi.com/exchange?from=${from}&to=${to}&q=1.0`;

describe("Exchange", () => {
  beforeEach(() => {
    fetchMock.mockReset();
  });

  it("renders a thing", () => {
    const queryClient = new QueryClient();
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should change the header text when user selects a new currency", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );
    expect(screen.getByText("Sell USD")).toBeInTheDocument();

    userEvent.click(screen.getByTestId("select-currency-from"));
    userEvent.click(screen.getByTestId("EUR"));

    expect(screen.getByText("Sell EUR")).toBeInTheDocument();
  });

  it("should change the subheader when the user selects a new currency", async () => {
    const fixture1 = "1.2";
    const fixture2 = "1.3";

    fetchMock.mock(getFetchUrl(ECurrency.Usd, ECurrency.Eur), fixture1);

    fetchMock.mock(getFetchUrl(ECurrency.Gbp, ECurrency.Eur), fixture2);

    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    expect(fetchMock).toHaveBeenCalledWith(
      getFetchUrl(ECurrency.Usd, ECurrency.Eur),
      fetchOptions
    );

    await waitFor(() => {
      expect(
        screen.getByText(`Market order • 1 USD = ${fixture1} EUR`)
      ).toBeInTheDocument();
    });

    userEvent.click(screen.getByTestId("select-currency-from"));
    userEvent.click(screen.getByTestId("GBP"));

    expect(fetchMock).toHaveBeenCalledWith(
      getFetchUrl(ECurrency.Gbp, ECurrency.Eur),
      fetchOptions
    );

    await waitFor(() => {
      expect(
        screen.getByText(`Market order • 1 GBP = ${fixture2} EUR`)
      ).toBeInTheDocument();
    });
  });

  it("should fetch a new rate every 10 seconds and update the subheader", async () => {
    const clock = FakeTimers.install();
    /**
     * Not possible to use Jest's Fake Timers with promises:
     * https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises/51132058#51132058
     *
     * Using Sinon's Fake Timers instead
     */
    //

    const fixture = "1.0";
    fetchMock.mock(getFetchUrl(ECurrency.Usd, ECurrency.Eur), fixture);

    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    expect(fetchMock.mock.calls.length).toEqual(1);
    await clock.tickAsync(10000);
    expect(fetchMock.mock.calls.length).toEqual(2);
    await clock.tickAsync(10000);
    expect(fetchMock.mock.calls.length).toEqual(3);

    clock.uninstall();
  });

  it("should have selected by default USD to EUR", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const selectFrom = screen.getByTestId("select-currency-from");
    const selectTo = screen.getByTestId("select-currency-to");

    expect(within(selectFrom).getByText("USD")).toBeInTheDocument();
    expect(within(selectTo).getByText("EUR")).toBeInTheDocument();
  });

  it("shoud have the inputs empty by default", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");
    const inputTo = screen.getByTestId("input-amount-to");

    expect(inputFrom).toHaveValue("");
    expect(inputTo).toHaveValue("");
  });

  it("should have the correct balances for each currency account", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    expect(
      within(screen.getByTestId("select-currency-from")).getByText("USD")
    ).toBeInTheDocument();

    const usdAccount = initialState.accounts.find(
      ({ id }) => id === ECurrency.Usd
    ) as IAccount;

    const eurAccount = initialState.accounts.find(
      ({ id }) => id === ECurrency.Eur
    ) as IAccount;

    expect(screen.getByTestId("account-balance-from").textContent).toEqual(
      `Balance: ${usdAccount.balance} ${usdAccount.id}`
    );

    expect(screen.getByTestId("account-balance-to").textContent).toEqual(
      `Balance: ${eurAccount.balance} ${eurAccount.id}`
    );
  });

  it("should accept only numbers in input with a maximum of 2 decimals", async () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");

    userEvent.type(inputFrom, "0.000");
    expect(inputFrom).toHaveValue("0.00");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "999999999.99");
    expect(inputFrom).toHaveValue("999999999.99");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "9999999999.99");
    expect(inputFrom).toHaveValue("999999999.99");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "123.12");
    expect(inputFrom).toHaveValue("123.12");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "45.137");
    expect(inputFrom).toHaveValue("45.13");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "0.1");
    expect(inputFrom).toHaveValue("0.1");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "0.123");
    expect(inputFrom).toHaveValue("0.12");
    userEvent.clear(inputFrom);

    userEvent.type(inputFrom, "+-e1");
    expect(inputFrom).toHaveValue("1");
    userEvent.clear(inputFrom);
  });

  it("should automatically fill to the second input when user types in the first input", () => {
    const fixture = "1.2";
    fetchMock.mock(getFetchUrl(ECurrency.Usd, ECurrency.Eur), fixture);

    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");
    const inputTo = screen.getByTestId("input-amount-to");

    userEvent.type(inputFrom, "123.12");

    waitFor(() => {
      expect(inputTo).toHaveValue("147.74"); // 123.12 * 1.2
    });
  });

  it("should update the Sell button label with the selected currency acronyms", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    expect(screen.getByTestId("sell-button")).toHaveTextContent(
      "Sell USD to EUR"
    );

    userEvent.click(screen.getByTestId("select-currency-from"));
    userEvent.click(screen.getByTestId("GBP"));

    userEvent.click(screen.getByTestId("select-currency-to"));
    userEvent.click(screen.getByTestId("USD"));

    expect(screen.getByTestId("sell-button")).toHaveTextContent(
      "Sell GBP to USD"
    );
  });

  it("should have the Sell button disabled by default when the first input does not have a value or the value is 0", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");

    expect(screen.getByTestId("sell-button")).toBeDisabled();

    userEvent.type(inputFrom, "0.");
    expect(screen.getByTestId("sell-button")).toBeDisabled();

    userEvent.type(inputFrom, "1");
    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeEnabled();
    });
  });

  it("should have the Sell button disabled when trying to exchange the same currency", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");

    userEvent.type(inputFrom, "100");

    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeEnabled();
    });

    userEvent.click(screen.getByTestId("select-currency-to"));
    userEvent.click(screen.getByTestId("USD"));

    expect(screen.getByTestId("sell-button")).toHaveTextContent(
      "Sell USD to USD"
    );

    expect(screen.getByTestId("sell-button")).toBeDisabled();
  });

  it("should disable the Sell button if the value in the first input exceeds the value in the balance for that currrency account", () => {
    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const usdAccount = initialState.accounts.find(
      ({ id }) => id === ECurrency.Usd
    ) as IAccount;

    const inputFrom = screen.getByTestId("input-amount-from");
    userEvent.type(inputFrom, usdAccount.balance);

    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeEnabled();
    });

    userEvent.type(
      inputFrom,
      String(currency(usdAccount.balance).add("0.01").value)
    );

    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeDisabled();
      expect(screen.getByText("exceed balance")).toBeInTheDocument();
    });
  });

  it("shoud convert with the correct rate when the user presses the Sell button and update the balances accordingly", () => {
    const rate = "0.1";
    const amount = "100";

    fetchMock.mock(getFetchUrl(ECurrency.Usd, ECurrency.Eur), rate);

    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const usdAccount = initialState.accounts.find(
      ({ id }) => id === ECurrency.Usd
    ) as IAccount;

    const eurAccount = initialState.accounts.find(
      ({ id }) => id === ECurrency.Eur
    ) as IAccount;

    const inputFrom = screen.getByTestId("input-amount-from");
    userEvent.type(inputFrom, amount);

    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeEnabled();

      userEvent.click(screen.getByTestId("sell-button"));

      expect(screen.getByTestId("account-balance-from").textContent).toEqual(
        `Balance: ${currency(usdAccount.balance).subtract(amount).value} ${
          usdAccount.id
        }`
      );

      expect(screen.getByTestId("account-balance-to").textContent).toEqual(
        `Balance: ${
          currency(eurAccount.balance).add(currency(amount).multiply(rate))
            .value
        } ${eurAccount.id}`
      );
    });
  });

  it("should reset the inputs when a currency exchange is made", () => {
    const rate = "0.1";
    const amount = "100";

    fetchMock.mock(getFetchUrl(ECurrency.Usd, ECurrency.Eur), rate);

    const queryClient = new QueryClient();
    render(
      <QueryClientProvider client={queryClient}>
        <Exchange />
      </QueryClientProvider>
    );

    const inputFrom = screen.getByTestId("input-amount-from");
    const inputTo = screen.getByTestId("input-amount-to");
    userEvent.type(inputFrom, amount);

    waitFor(() => {
      expect(screen.getByTestId("sell-button")).toBeEnabled();

      userEvent.click(screen.getByTestId("sell-button"));

      expect(inputFrom).toHaveValue("");
      expect(inputTo).toHaveValue("");
    });
  });
});
