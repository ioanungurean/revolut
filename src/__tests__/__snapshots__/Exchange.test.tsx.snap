// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange renders a thing 1`] = `
.c5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 140px;
}

.c6 {
  font-size: 18px;
  line-height: 38px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 38px;
  padding: 4px 10px;
  cursor: pointer;
  border: 1px solid;
  border: 1px solid #767676;
  border-radius: 4px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c7 {
  font-size: 12px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  font-size: 24px;
  font-weight: bold;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  font-size: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 12px;
  color: #2684ff;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 12px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 360px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  text-align: right;
  width: 140px;
  padding: 4px;
}

.c9 {
  font-size: 12px;
  padding-top: 6px;
}

.c10 {
  font-size: 16px;
  width: 360px;
  padding: 12px 36px;
  cursor: pointer;
  text-align: center;
  color: white;
  border: none;
  background-color: #2684ff;
}

.c10:hover {
  background-color: #0c50a6;
}

.c10:disabled {
  cursor: default;
  background-color: #c8c8c8;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Sell 
      USD
    </div>
    <div
      class="c2"
    >
      Market order • 1 
      USD
       = 
       
      EUR
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
            data-testid="select-currency-from"
            id="select-currency-from"
          >
            <span>
              USD
            </span>
            <span
              class="c7"
            >
              ▼
            </span>
          </div>
        </div>
        <input
          class="c8"
          data-testid="input-amount-from"
          placeholder="0.00 USD"
          value=""
        />
      </div>
      <div
        class="c4"
      >
        <span
          class="c9"
          data-testid="account-balance-from"
        >
          Balance: 
          874.12
           
          USD
        </span>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="c6"
            data-testid="select-currency-to"
            id="select-currency-to"
          >
            <span>
              EUR
            </span>
            <span
              class="c7"
            >
              ▼
            </span>
          </div>
        </div>
        <input
          class="c8"
          data-testid="input-amount-to"
          placeholder="0.00 EUR"
          value=""
        />
      </div>
      <div
        class="c4"
      >
        <span
          class="c9"
          data-testid="account-balance-to"
        >
          Balance: 
          45.04
           
          EUR
        </span>
      </div>
    </div>
    <button
      class="c10"
      data-testid="sell-button"
      disabled=""
    >
      Sell 
      USD
       to 
      EUR
    </button>
  </div>
</div>
`;
